File: backend/app/schemas/gamification.py
"""Pydantic schemas for leaderboard / skill tree / profile (Phase 18)."""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

TierLiteral = Literal["diamond", "platinum", "gold", "silver", "bronze"]
UnlockStateLiteral = Literal["locked", "unlocked", "mastered"]


# ── Leaderboard ──

class LeaderboardRowResponse(BaseModel):
    rank: int
    student_id: str
    name: str
    total_score: float
    academic: float
    skill: float
    interview: float
    placement: float
    tier: TierLiteral
    xp_total: int
    current_level: int
    is_current_user: bool


class LeaderboardResponse(BaseModel):
    rows: list[LeaderboardRowResponse]
    my_row: LeaderboardRowResponse | None = None
    total_students: int


# ── Skill Tree ──

class SkillNodeStateResponse(BaseModel):
    name: str
    domain: str
    description: str
    difficulty_tier: int
    mastery: float
    state: UnlockStateLiteral


class DomainProgressResponse(BaseModel):
    name: str
    total_skills: int
    mastered_count: int
    unlocked_count: int
    progress_percent: float
    skills: list[SkillNodeStateResponse]


class SkillTreeResponse(BaseModel):
    domains: list[DomainProgressResponse]
    mastered_count: int
    total_count: int
    overall_percent: float


# ── Profile ──

class BadgeResponse(BaseModel):
    id: str
    badge_type: str
    name: str
    description: str | None
    icon: str | None
    earned_at: str


class PublicProfileResponse(BaseModel):
    user_id: uuid.UUID
    name: str
    role: str
    branch: str | None
    semester: int | None
    cgpa: float | None
    github_url: str | None
    linkedin_url: str | None
    skills: list[str]
    target_companies: list[str]
    target_role: str | None
    xp_total: int
    current_level: int
    streak_days: int
    campus_iq_score: float
    rank: int | None
    tier: TierLiteral
    academic_pillar: float
    skill_pillar: float
    interview_pillar: float
    placement_pillar: float
    badges: list[BadgeResponse]
    member_since: datetime


class ProfileUpdateRequest(BaseModel):
    branch: str | None = Field(None, max_length=100)
    semester: int | None = Field(None, ge=1, le=12)
    cgpa: float | None = Field(None, ge=0, le=10)
    github_url: str | None = Field(None, max_length=500)
    linkedin_url: str | None = Field(None, max_length=500)
    skills: list[str] | None = None
    target_companies: list[str] | None = None
    target_role: str | None = Field(None, max_length=255)

File: frontend/src/types/gamification.ts
/**
 * Gamification + profile types — mirror backend Pydantic schemas in
 * backend/app/schemas/gamification.py
 */

export type Tier = 'diamond' | 'platinum' | 'gold' | 'silver' | 'bronze'
export type UnlockState = 'locked' | 'unlocked' | 'mastered'

export interface LeaderboardRowResponse {
  rank: number
  student_id: string
  name: string
  total_score: number
  academic: number
  skill: number
  interview: number
  placement: number
  tier: Tier
  xp_total: number
  current_level: number
  is_current_user: boolean
}

export interface LeaderboardResponse {
  rows: LeaderboardRowResponse[]
  my_row: LeaderboardRowResponse | null
  total_students: number
}

export interface SkillNodeStateResponse {
  name: string
  domain: string
  description: string
  difficulty_tier: number
  mastery: number
  state: UnlockState
}

export interface DomainProgressResponse {
  name: string
  total_skills: number
  mastered_count: number
  unlocked_count: number
  progress_percent: number
  skills: SkillNodeStateResponse[]
}

export interface SkillTreeResponse {
  domains: DomainProgressResponse[]
  mastered_count: number
  total_count: number
  overall_percent: number
}

export interface BadgeResponse {
  id: string
  badge_type: string
  name: string
  description: string | null
  icon: string | null
  earned_at: string
}

export interface PublicProfileResponse {
  user_id: string
  name: string
  role: string
  branch: string | null
  semester: number | null
  cgpa: number | null
  github_url: string | null
  linkedin_url: string | null
  skills: string[]
  target_companies: string[]
  target_role: string | null
  xp_total: number
  current_level: number
  streak_days: number
  campus_iq_score: number
  rank: number | null
  tier: Tier
  academic_pillar: number
  skill_pillar: number
  interview_pillar: number
  placement_pillar: number
  badges: BadgeResponse[]
  member_since: string
}

export interface ProfileUpdateRequest {
  branch?: string | null
  semester?: number | null
  cgpa?: number | null
  github_url?: string | null
  linkedin_url?: string | null
  skills?: string[]
  target_companies?: string[]
  target_role?: string | null
}
